duet/typing.py

```diff
@@ -37,10 +37,19 @@ def sync(f: Callable[..., Awaitable[T]]) -> Callable[..., T]:
     functions wrapped by duet.sync.
     """
     func_type = get_proper_type(ctx.arg_types[0][0])
-    if not isinstance(func_type, CallableType):
+    if not isinstance(func_type, (CallableType, Overloaded)):
         ctx.api.msg.fail(f"expected Callable[..., Awaitable[T]], got {func_type}", ctx.context)
         return ctx.default_return_type
 
+    if isinstance(func_type, CallableType):
+        return modify_callable(func_type, ctx)
+
+    # func_type is overloaded
+    overloaded_callables = [cast(CallableType, modify_callable(ft, ctx)) for ft in func_type.items]
```

Comment:
The issue is the return statement on line 60.

Reply:
Ok, but then in that case I'd prefer to provide an explicit check here rather than casting. For example, we could do something like:
```python
overloaded_callables: List[CallableType] = []
for ft in func_type.items:
    overload_type = modify_callable(ft, ctx)
    if not isinstance(overload_type, CallableType):
        raise TypeError(...)
    overloaded_callables.append(overload_type)
```

Another possibility would be to have a separate function similar to `modify_callable` (or a flag on that function) that will fail if the return type is not a coroutine, rather than falling back to `ctx.default_return_type`, which doesn't really make sense to return for an individual overload case.

Reply:
I am not sure if raising errors here is appropriate, it leads to a cryptic message if it's ever raised when typechecking
```
duet/api.py:152: error: INTERNAL ERROR -- Please try using mypy master on Github:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
If this issue continues with mypy master, please report a bug at https://github.com/python/mypy/issues
version: 0.941
duet/api.py:152: : note: please use --show-traceback to print a traceback when reporting a bug
```

What do you think of having `modify_callable` just return `func_type` if it has the wrong return type? I don't really understand what role the return even plays if we have called `ctx.api.msg.fail`

Reply:
PTAL at the latest commit

Reply:
Sorry, I should have said to call `ctx.api.msg.fail` rather than raise an error, e.g.:
```python
overloaded_callables: List[CallableType] = []
for ft in func_type.items:
    overload_type = modify_callable(ft, ctx)
    if not isinstance(overload_type, CallableType):
        ctx.api.msg.fail(f"expected overloaded type to be callable, got {overload_type}", ctx.context)
        return ctx.default_return_type
    overloaded_callables.append(overload_type)
```
I don't think we want to return `func_type` because this won't include some type inference that mypy does.

Reply:
Done, thanks!